src/databricks/labs/ucx/source_code/notebook.py

```diff
@@ -0,0 +1,331 @@
+
+    def __init__(self, *args):
+        super().__init__()
+        self._language = args[0]
+        self._magic_name = args[1]
+        self._comment_prefix = args[2]
+        self._new_cell = args[3]
+
+    @property
+    def language(self) -> Language:
+        return self._language
+
+    @property
+    def magic_name(self) -> str:
+        return self._magic_name
+
+    @property
+    def comment_prefix(self) -> str:
+        return self._comment_prefix
+
+    @classmethod
+    def of_language(cls, language: Language) -> CellLanguage:
+        return next((cl for cl in CellLanguage if cl.language == language))
+
+    @classmethod
+    def of_magic_name(cls, magic_name: str) -> CellLanguage | None:
+        return next((cl for cl in CellLanguage if magic_name.startswith(cl.magic_name)), None)
+
+    def read_cell_language(self, lines: list[str]) -> CellLanguage | None:
+        magic_prefix = f'{self.comment_prefix}{MAGIC_PREFIX}'
+        magic_language_prefix = f'{magic_prefix}{LANGUAGE_PREFIX}'
+        for line in lines:
+            # if we find a non-comment then we're done
+            if not line.startswith(self.comment_prefix):
+                return None
+            # if it's not a magic comment, skip it
+            if not line.startswith(magic_prefix):
+                continue
+            if line.startswith(magic_language_prefix):
+                line = line[len(magic_language_prefix) :]
+                return CellLanguage.of_magic_name(line.strip())
+            return None
+        return None
+
+    def new_cell(self, source: str) -> Cell:
+        return self._new_cell(source)
+
+
+def extract_cells(source: str, default_language: CellLanguage) -> list[Cell] | None:
+    lines = source.split('\n')
+    header = f"{default_language.comment_prefix}{NOTEBOOK_HEADER}"
+    if not lines[0].startswith(header):
+        raise ValueError("Not a Databricks notebook source!")
+
+    def make_cell(lines_: list[str]):
+        # trim leading blank lines
+        while len(lines_) > 0 and len(lines_[0]) == 0:
+            lines_.pop(0)
+        # trim trailing blank lines
+        while len(lines_) > 0 and len(lines_[-1]) == 0:
+            lines_.pop(-1)
+        cell_language = default_language.read_cell_language(lines_)
+        if cell_language is None:
+            cell_language = default_language
+        cell_source = '\n'.join(lines_)
+        return cell_language.new_cell(cell_source)
+
+    cells = []
+    cell_lines: list[str] = []
+    separator = f"{default_language.comment_prefix}{CELL_SEPARATOR}"
+    for i in range(1, len(lines)):
+        line = lines[i].strip()
+        if line.startswith(separator):
+            cell = make_cell(cell_lines)
+            cells.append(cell)
+            cell_lines = []
+        else:
+            cell_lines.append(lines[i])
+    if len(cell_lines) > 0:
+        cell = make_cell(cell_lines)
+        cells.append(cell)
+
+    return cells
+
+
+class NotebookDependencyGraph:
```

Comment:
```suggestion
class DependencyGraph:
```
might be less redundant this way - we're in `notebooks.py` anyway

Reply:
Yeah I was hesitant about this when writing it... done